src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectSystem/FileWatcherBasedRazorProjectInfoDriver.cs

```diff
@@ -0,0 +1,180 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.ProjectSystem;
+using Microsoft.AspNetCore.Razor.Utilities;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.Logging;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor.Utilities;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.ProjectSystem;
+
+internal partial class FileWatcherBasedRazorProjectInfoDriver : AbstractRazorProjectInfoDriver
+{
+    private enum ChangeKind { AddOrUpdate, Remove }
+
+    private static readonly ImmutableArray<string> s_ignoredDirectories =
+    [
+        "node_modules",
+        "bin",
+        ".vs",
+    ];
+
+    private readonly IWorkspaceRootPathProvider _workspaceRootPathProvider;
+    private readonly LanguageServerFeatureOptions _options;
+    private readonly ILogger _logger;
+
+    private readonly AsyncBatchingWorkQueue<(string FilePath, ChangeKind Kind)> _workQueue;
+
+    private FileSystemWatcher? _fileSystemWatcher;
+
+    public FileWatcherBasedRazorProjectInfoDriver(
+        IWorkspaceRootPathProvider workspaceRootPathProvider,
+        LanguageServerFeatureOptions options,
+        ILoggerFactory loggerFactory,
+        TimeSpan? delay = null)
+        : base(loggerFactory, delay)
+    {
+        _workspaceRootPathProvider = workspaceRootPathProvider;
+        _options = options;
+        _logger = loggerFactory.GetOrCreateLogger<FileWatcherBasedRazorProjectInfoDriver>();
+
+        _workQueue = new(delay ?? DefaultDelay, ProcessBatchAsync, DisposalToken);
+
+        // Dispose our FileSystemWatcher when the driver is disposed.
+        DisposalToken.Register(() =>
+        {
+            _fileSystemWatcher?.Dispose();
+            _fileSystemWatcher = null;
+        });
+    }
+
+    protected override async Task InitializeAsync(CancellationToken cancellationToken)
+    {
+        var workspaceDirectoryPath = await _workspaceRootPathProvider.GetRootPathAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
I think its a bit odd that we start this up immediately, and then immediately wait for `IOnInitialize` indirectly through the workspace path provider.  I wonder if this self-initialization should/could be replaced with this implementing `IOnIntiailized`?

Reply:
Initially, I tried using IOnInitialized and ran into problems with ordering. Since IOnInitialized offers no guarantees about what order services will be initialized in, it's easy to get into trouble when there are dependencies between services that are all marked for initialization.

Reply:
Note that there's no such interface as `IOnInitialize`. I just want to be sure that we're getting the terminology right. So, I'm stating all of this to make my logic clear:

1. There are two LSP messages: [`"initialize"`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#initialize) and [`"initialized"`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#initialized). The first one happens before the second one.
2. The first LSP message sent is `"initialize"`. When received by `RazorInitializeEndpoint`, `CapbilitiesManager.SetInitializeParams(...)` is called, making the `RootPath` and `RootUri` values available.
3. The `IWorkspaceRootPathProvider.GetRootAsync(...)` implementation on `CapabilitiesManager` spins until the `InitializeParams` become available. Once available, it returns the root path value.
4. The next LSP message should be `"initialized"`. When received by `RazorInitializedEndpoint`, it loops through all `IOnInitialized` LSP services and calls `OnInitializedAsync`.

In my opinion, it's OK to await `IWorkspaceRootPathProvider.GetRootAsync(...)` here. This happens _before_ `"initialized"` and allows `FileWatcherBasedRazorProjectInfoDriver` to start loading projects early, so it's slightly more likely that the server will be ready to provide project information for the first LSP message that needs it.

Reply:
> Note that there's no such interface as `IOnInitialize`

I think that makes all of my points invalid. I'm not sure where I got the idea from.

Reply:
No worries! Another reason I clarified was to make sure that it wasn't just a typo. It could have just been `IOnInitialized` with the 'd' left off. 😄